src/cla/tcp.rs

```diff
@@ -0,0 +1,511 @@
+    timeout: u16,
+}
+
+struct TcpSender {
+    stream_writer: OwnedWriteHalf,
+    session_channel: Receiver<TcpClPacket>,
+    timeout: u16,
+}
+
+impl TcpReceiver {
+    async fn run(mut self) {
+        let mut interval = time::interval(Duration::from_secs(self.timeout.into()));
+        loop {
+            tokio::select! {
+                    parsed_packet = TcpClPacket::parse(&mut self.stream_reader) => {
+                        match parsed_packet {
+                            Ok(packet) => {
+                                debug!("Received and successfully parsed packet");
+                                self.send_packet(packet).await;
+                            }
+                            Err(err) => {
+                                error!("Failed parsing package: {:?}", err);
+                            }
+                        }
+                    }
+                    _ = interval.tick() => {
+                        self.send_packet(TcpClPacket::SessTerm(SessTermData {
+                            flags: SessTermFlags::empty(),
+                            reason: SessTermReasonCode::IdleTimeout,
+                        })).await;
+                }
+            }
+        }
+    }
+    async fn send_packet(&mut self, packet: TcpClPacket) {
+        if let Err(err) = self.session_channel.send(packet).await {
+            error!("Error while sending via internal channel: {}", err);
+        }
+    }
+}
+
+impl TcpSender {
+    async fn run(mut self) {
+        let mut interval = time::interval(Duration::from_secs(self.timeout.into()));
+        loop {
+            tokio::select! {
+                packet = self.session_channel.recv() => {
+                    if let Some(packet) = packet {
+                       self.send_packet(&packet).await;
+                       if let TcpClPacket::SessTerm(_) = packet {
+                           //breaks loop, tasks finished, dropped sender, connection closed
+                           break;
+                       }
+                    }
+                }
+                _ = interval.tick() => {
+                    self.send_packet(&TcpClPacket::KeepAlive).await;
+                }
+            }
+        }
+    }
+    async fn send_packet(&mut self, packet: &TcpClPacket) {
+        if let Err(err) = packet.serialize(&mut self.stream_writer).await {
+            error!("Error while serializing packet: {}", err);
+        }
+        if let Err(err) = self.stream_writer.flush().await {
+            error!("Error while flashing tcp sending queue: {}", err);
+        }
+    }
+}
+
+struct Connection {
+    stream: TcpStream,
+    addr: SocketAddr,
+}
+
+impl Connection {
+    async fn negotiate_session(&mut self) -> anyhow::Result<(SessInitData, SessInitData)> {
+        let mut sess_init_data = SessInitData {
+            keepalive: 30,
+            segment_mru: 64000,
+            transfer_mru: 64000,
+            node_id: "test2".into(),
```

Comment:
hardcoded `node_id` string? shouldn't this be taken from the currently running instance?